api/src/main/java/org/openmrs/obs/handler/AbstractHandler.java

```diff
@@ -59,46 +61,30 @@ public AbstractHandler() {
 	 * @return File that the complex data should be written to
 	 */
 	public File getOutputFileToWrite(Obs obs) throws IOException {
-		// Get the title and remove the extension.
+		// Get the title and the extension
 		String title = obs.getComplexData().getTitle();
-		String extension = "." + getExtension(title);
-		
-		// If getExtension returns the title, there was no extension
-		if (getExtension(title).equals(title)) {
-			extension = "";
-		}
+		String extension = FilenameUtils.getExtension(title);
 		
 		File dir = OpenmrsUtil.getDirectoryInApplicationDataDirectory(Context.getAdministrationService().getGlobalProperty(
 		    OpenmrsConstants.GLOBAL_PROPERTY_COMPLEX_OBS_DIR));
-		File outputfile;
+		File outputfile = null;
 		
-		// Get the output stream
+		// Generate the filename with the Obs' UUID
+		String uuid = obs.getUuid();
 		if (null == title) {
 			String now = longfmt.format(new Date());
-			outputfile = new File(dir, now);
+			outputfile = new File(dir, now + '_' + uuid);
 		} else {
-			title = title.replace(extension, "");
+			// If the extension is no empty string, it exists and we can add a dot in front for later use
+			if (!(extension.equals(""))) {
+				extension = "." + extension;
+			}
 			
-			outputfile = new File(dir, title + extension);
-		}
-		
-		int i = 0;
-		String tmp;
-		
-		// If the Obs does not exist, but the File does, append a two-digit
-		// count number to the filename and save it.
-		while (obs.getObsId() == null && outputfile.exists() && i < 100) {
-			// Remove the extension from the filename.
-			tmp = String.valueOf(outputfile.getAbsolutePath().replace(extension, ""));
-			// Append two-digit count number to the filename.
-			String filename = (i < 1) ? tmp + "_" + nf.format(Integer.valueOf(++i)) : tmp.replace(nf.format(Integer
-			        .valueOf(i)), nf.format(Integer.valueOf(++i)));
-			// Append the extension to the filename
-			outputfile = new File(filename + extension);
+			String titleWithoutExtension = title.replace(extension, "");
```

Comment:
I would have bet there an API to do things like that in (Apache's) `FilenameUtils`.
Just in case the actual string `extension` exists multiple time in the file. Imagine a:
```
my_txt_file.txt
```

Reply:
This particular example would not be a problem I think, as I did add a dot to the extension String beforehand. But you're right in that there could probably be malformed names an API could maybe handle better. I will look for an appropriate method.